File: 199-binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:

        if not root:
            return []

        Q = collections.deque([root])
        res = []

        while Q:

            cursize = len(Q)

            for i in range(cursize):

                node = Q.popleft()

                if i == cursize -1:
                    res.append(node.val)
                
                if node.left:
                    Q.append(node.left)

                if node.right:
                    Q.append(node.right)
            
        return res

        

File: 199-binary-tree-right-side-view/binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {

        Queue <TreeNode> Q = new LinkedList<>();
        List <Integer> arr = new ArrayList<>();

        if(root != null){
            Q.add(root);
        }

        while(!Q.isEmpty() ){

            int cur_size = Q.size();

            for(int i=0;i<cur_size;i++){

                TreeNode node = Q.poll();

                if(i == cur_size -1){
                    arr.add(node.val);
                }

                if(node.left != null){
                    Q.add(node.left);
                }

                if(node.right != null){
                    Q.add(node.right);
                }


            }

        }

        return arr;
        
    }
}